### ISSUE TITLE
DOC: freeze old whatsnew notes?

### ISSUE THREAD

**Issue by jorisvandenbossche** [pandas-dev/pandas#6856]
A problem with using the ipython directory for the whatsnew/release parts of the docs (so that all code examples are run during the doc build), is that after a while when things change/get deprecated, these pages don't build correctly anymore, generating warnings/exceptions.

An example: http://pandas-docs.github.io/pandas-docs-travis/whatsnew.html#hdfstore, with the `unique` attribute that was removed recently in #6813

I don't think we should start updating the older whatsnew docs (they are a kind of historical overview), but we should do something as those doc build warnings are annoying. A proposal:
- after a release (or after some time), we run whatsnew/release once with the pandas version of that release, and replace all `ipython` directory blocks to literal `code-blocks` with output.
- we maybe add a warning at the top of each whatsnew that the code examples can be outdated

An alternative is to add `:okexcept:` to those ipython code-blocks; this would solve the warnings during the doc build, but not the 'faulty' output in the docs itself.

**Comment by jreback** [comment]
@jorisvandenbossche IIRC you fixed this?

**Comment by jreback** [comment]
update on this.

These older ``whatsnew`` *are* building pretty well with a couple of minor exceptions.

However, it would be nice to changes examples from using ipython -> code-blocks to:

- make this faster
- make less error prone

**Comment by jorisvandenbossche** [comment]
This would indeed be nice to do (at least for the very old ones, for the relatively recent ones it is still a good way sometimes to capture errors)

I at some point looked into a kind of 'text' sphinx builder that would output what we want, but never really figured it out. I think it should be somehow possible (eg numpydoc also emits the intermediate generated source code that is never saved in a file to build the actual docstring page when using very verbose mode)

**Comment by avinashpancham** [comment]
While working on #36838 I got exactly this issue. If I understand it correctly we want to remove the `ipython` directory in old whatsnew notes and replace it with `code-blocks`. If we still want to do that, then I can work on this PR

**Comment by jorisvandenbossche** [comment]
@avinashpancham sorry for the late reply, but yes, I think that's still something we want to do. (ideally we would have a way to do this automatically, though, I think. In *theory* it should be possible with some sphinx / ipython directive hacks I think)

**Comment by fangchenli** [comment]
See #41464 for how to freeze old whatsnew notes. It's very simple but labor-intensive.

**Comment by jorisvandenbossche** [comment]
Yeah, if we want to do this for the later (and much bigger) whatsnew files as well, I think some automatic method (using sphinx / ipython directive) would be better

**Comment by derrick-anderson** [comment]
I'm digging into older "easy" tickets and hoping to brush off my contribution skills. 

I see two issues here:
1. Freeze "old" `whatsnew` notes in order to prevent syntax errors.
  a.What do we consider old?  Anything not the most recent? 
2. Find a way to programmatically do this. 

For 1. it's definitely labor intensive but not undoable, I'm happy to help tackle.
For 2 however this one scratches an itch i've had for a while TOOLING.  I'll start to investigate some patterns.  I like the idea of potentially overriding a sphynx directive.  

Would we want to split this into two issues?  I can submit a new one for the tooling enhancement.  Then we could use this one to run a check list of the outstanding docs to convert so that this can still be closed with a goal in mind.

**Comment by derrick-anderson** [comment]
Hi Everyone! 

I have been doing a bunch of digging on this topic.  I've come down on [this solution](https://stackoverflow.com/a/63460377) that allows you to lock-in a specific version of a library for build time.  This would require the builder to install all the possible versions of pandas on the build machine. That is possible with a requires file and then we would include a global import for that specific version at the top of the page for that version.  There's also an option to drop that work into the `execlines` for that page, but then a whole new Sphinx shell would need to be spun up for each page, probably making those builds even longer.

However, over time I am agreeing with the idea that there likely is no real value in re-running all these old code snippets during build time.  The only real value I can see is when a change is back-ported then the value in a static code block won't change.  This however is a rare occurrence and the docs being wrong should be significantly less of an issue than trying to wedge something into the build machine and putting heavier onus on contributors. 

I'm going to get started converting all the old iPython directive blocks into `code-block::ipython` and drop in the appropriate output from that version of pandas.  We should probably note somewhere to use that directive for docs after that.

Let me know if you have other ideas! 

-Derrick

**Comment by TheMellyBee** [comment]
Can this be closed?

**Comment by jorisvandenbossche** [comment]
I think this would still be nice to do

**Comment by TheMellyBee** [comment]
Sounds good! This looks like it needs some refactoring, updating, and just basic maintenance. Would you like some help on it?

**Comment by rahuindia9** [comment]
Hi, I want to take this issue

**Comment by Lk-ux** [comment]
take
